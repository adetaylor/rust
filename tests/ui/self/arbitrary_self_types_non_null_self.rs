// run-pass

#![feature(receiver_trait)]

use std::ptr::NonNull;

struct Foo(usize);

impl Foo {
  fn foo(self: &NonNull<Self>) -> i32 { 3 }
  fn bar(self: &mut NonNull<Self>) -> i32 { 3 }
  fn baz(self: NonNull<Self>) -> i32 { 3 }
  fn as_mut(self: &mut NonNull<Self>) -> i32 { 4 }
}

fn main() {
  let mut foo = Foo(3);
  let mut ptr = std::ptr::NonNull::new(&mut foo as *mut Foo).unwrap();
  assert_eq!(ptr.as_mut(), 4);
  //~^ ambiguous targets
  assert_eq!(ptr.foo(), 3);
  assert_eq!(ptr.bar(), 3);
  assert_eq!(ptr.baz(), 3);
}
