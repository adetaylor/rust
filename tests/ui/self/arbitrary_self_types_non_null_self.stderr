warning: NonNull::<T>::as_mut and Foo::as_mut are ambiguous targets of this method call; explicitly choose one
  --> $DIR/arbitrary_self_types_non_null_self.rs:19:18
   |
LL |   fn as_mut(self: &mut NonNull<Self>) -> i32 { 4 }
   |   ------------------------------------------ because we'll call this method
...
LL |   assert_eq!(ptr.as_mut(), 4);
   |                  ^^^^^^
   |
   = note: the method in crate core won't be called
   = note: `#[warn(shadowed_method_picked)]` on by default
help: call as a function not a method
   |
LL |   assert_eq!(ptr.Foo::as_mut(FOO, ARGUMENTS)(), 4);
   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~
help: call as a function not a method
   |
LL |   assert_eq!(ptr.NonNull::<T>::as_mut(FOO, ARGUMENTS)(), 4);
   |                  ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

warning: 1 warning emitted

